### ISSUE TITLE
Rewriting file on one node is not always noticed on another - with file-cache-timeout-in-seconds=0

### ISSUE THREAD

**Issue by cris-b** [Azure/azure-storage-fuse#668]
### Which version of the blobfuse was used?

1.3.8, 1.4.1

### Which OS (please include version) are you using?

Ubuntu 18.04 (AKS)

### What problem was encountered?

When running blobfuse on two nodes, with the cache timeout set to zero, I observed that sometimes one node does not get the correct file content, in this pattern:

node 1:  echo hello > helloworld
node 2:  cat helloworld
hello
node 1: echo hello2 > helloworld
node2: cat helloworld
hello
node 1: echo hello3 > helloworld
node2: cat helloworld
hello

So this is wrong, and no amount of waiting updates the file.  What does update the file on node 2 is appending:

node1: echo hello3 >> helloworld
node2: cat helloworld
hello3
hello3

So appending always works and we get the fresh copy of the file instantly on the other node.  But rewriting the file a second time seems to cause things to become out of sync.

### Have you found a mitigation/solution?

No.

### If relevant, please share your mount command.

--file-cache-timeout-in-seconds=0 --use-adls=true

**Comment by vibhansa-msft** [comment]
For small files, kernel will also cache the contents and the read operation will be served from the kernel itself, it will not even land to blobfuse. Along with file-cache-timeout=0 try cleaning up kernel cache manually and validate.

**Comment by vibhansa-msft** [comment]
On node2 where you are doing cat use this before cat and see : sysctl -w vm.drop_caches=3
Also file-cache-timeout shall be 0 on all three nodes. Blobfuse caches attributes of file as well so you can disable that by removing --attr-cache cli param. However this will reduce performance a bit.

**Comment by cris-b** [comment]
The docs suggest that attr-cache is false by default?

setting -o direct_io on the mount seems to resolve this problem, as this stops the kernel caching.  But from the auto_cache parameter which to an extent (supposedly) controls this, it mentions that the cache refresh is based on file size and mtime.  But blobfuse only seems to be regarding the file size.  mtime on the file is updated

.

**Comment by vibhansa-msft** [comment]
- Blobfuse refresh the cache based on file-cache-timeout and its counted from last downloaded, and eviction is only possible after all handles are closed for that file. We can not rely on last modified time as it reflects the actual time when file was last updated in container, and it can be days old as well. 
- Attr-cache is disabled by default.
- If kernel is caching the content then blobfuse will not even be aware that a read is done on file, as the read call will return back from kernel. Hence in this case there is nothing blobfuse can do. For this case kernel cache needs to be wiped out through parameter or external cli as I suggested earlier.

**Comment by cris-b** [comment]
Yes I completely agree that this is not a problem in the blobfuse part per se.

It's a problem because the version of userspace part of fuse that is used (libfuse2) does not support the option which would make this work. 

See this useful page for what kernel contains what api:
https://john-millikin.com/the-fuse-protocol#versions

libfuse2 advertises support for up to version 7.19 of the kernel api.  7.20 introduced several new options, including FUSE_AUTO_INVAL_DATA, which means the kernel cache is invalided on file change OR mtime.

If I patch my fuse2 with the attached patch (which makes it advertise support for 7.20 and pass through the option), then add 

`conn->want |= FUSE_CAP_AUTO_INVAL_DATA;`

in `azs_init()`

I get the working behaviour without having to go via direct_io route.

So you see this is not a new option (its been in kernel for 9 years - Sep 2012), but libfuse2 is too old to have advertised supported for it, the last major version (2.9) having been released in April 2012.

fuse 2 patch (FYI - this is the only related code to this option in fuse3):

```
diff -urw fuse-2.9.9/include/fuse_kernel.h fuse-2.9.9.new/include/fuse_kernel.h
--- fuse-2.9.9/include/fuse_kernel.h    2019-01-04 13:33:33.000000000 +0000
+++ fuse-2.9.9.new/include/fuse_kernel.h        2021-11-14 08:41:24.769923620 +0000
@@ -119,7 +119,7 @@
 #define FUSE_KERNEL_VERSION 7
 
 /** Minor version number of this interface */
-#define FUSE_KERNEL_MINOR_VERSION 19
+#define FUSE_KERNEL_MINOR_VERSION 20
 
 /** The node ID of the root inode */
 #define FUSE_ROOT_ID 1
diff -urw fuse-2.9.9/lib/fuse_lowlevel.c fuse-2.9.9.new/lib/fuse_lowlevel.c
--- fuse-2.9.9/lib/fuse_lowlevel.c      2019-01-04 13:33:33.000000000 +0000
+++ fuse-2.9.9.new/lib/fuse_lowlevel.c  2021-11-14 08:28:38.341496816 +0000
@@ -33,6 +33,9 @@
 #define F_SETPIPE_SZ   (F_LINUX_SPECIFIC_BASE + 7)
 #endif
 
+#define FUSE_AUTO_INVAL_DATA ( 1 << 12 )
+#define FUSE_CAP_AUTO_INVAL_DATA (1 << 12 )
+
 
 #define PARAM(inarg) (((char *)(inarg)) + sizeof(*(inarg)))
 #define OFFSET_MAX 0x7fffffffffffffffLL
@@ -1782,6 +1785,8 @@
                        f->conn.capable |= FUSE_CAP_POSIX_LOCKS;
                if (arg->flags & FUSE_ATOMIC_O_TRUNC)
                        f->conn.capable |= FUSE_CAP_ATOMIC_O_TRUNC;
+                if (arg->flags & FUSE_AUTO_INVAL_DATA)
+                        f->conn.capable |= FUSE_CAP_AUTO_INVAL_DATA;
                if (arg->flags & FUSE_EXPORT_SUPPORT)
                        f->conn.capable |= FUSE_CAP_EXPORT_SUPPORT;
                if (arg->flags & FUSE_BIG_WRITES)
@@ -1856,6 +1861,9 @@
                outarg.flags |= FUSE_DONT_MASK;
        if (f->conn.want & FUSE_CAP_FLOCK_LOCKS)
                outarg.flags |= FUSE_FLOCK_LOCKS;
+        if (f->conn.want & FUSE_CAP_AUTO_INVAL_DATA)
+                outarg.flags |= FUSE_AUTO_INVAL_DATA;
+
        outarg.max_readahead = f->conn.max_readahead;
        outarg.max_write = f->conn.max_write;
        if (f->conn.proto_minor >= 13) {

```

**Comment by vibhansa-msft** [comment]
Thanks @cris-b for a such a detailed analysis. I understood your point that libfuse2 (2.9) is not having any option for INVAL_DATA by default and fuse3 has it (also support for other >=7.20 kernel options maybe).

At this point I do not see much can be done on blobfuse side (unless fuse2 has that option or blobfuse move to fuse3). You have to rely on a private build for time being. 

Let me know if you have any suggestion for blobfuse on this front, which might help with your use case.

**Comment by vibhansa-msft** [comment]
We will not be adding fuse3 support in blobfuse. We are working on the next generation of blobfuse and that will natively support fuse3 along with the kernel caching flag. Preview version of the same is already posted on github.

**Comment by andyzhangx** [comment]
@vibhansa-msft so this issue is not fixed on v1.4.4 or bobfuse2?

**Comment by vibhansa-msft** [comment]
No in Blobfuse 1.x we do not have fuse option to invalidate the kernel cache on attribute changes. In fuse3 we can do that and hence we have integrated this part in Blobfuse2.
For V1 only option user has is to manually wipe out kernel page cache.
